cloudevents/sdk/http_events.py

```diff
@@ -11,183 +11,157 @@
 #    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 #    License for the specific language governing permissions and limitations
 #    under the License.
-import copy
-
-import io
 
+import datetime
 import json
 import typing
+import uuid
 
 from cloudevents.sdk import converters
 from cloudevents.sdk import marshaller
+from cloudevents.sdk import types
 
 from cloudevents.sdk.event import v03, v1
 
+_marshaller_by_format = {
+    converters.TypeStructured: lambda x: x,
+    converters.TypeBinary: json.dumps,
+}
+_required_by_version = {
+    "1.0": v1.Event._ce_required_fields,
+    "0.3": v03.Event._ce_required_fields,
+}
+_obj_by_version = {"1.0": v1.Event, "0.3": v03.Event}
+
+
+def _json_or_string(content: typing.Union[str, bytes]):
+    if len(content) == 0:
+        return None
+    try:
+        return json.loads(content)
+    except json.JSONDecodeError:
+        return content
+
 
 class CloudEvent():
     """
     Python-friendly cloudevent class supporting v1 events
-    Currently only supports binary content mode CloudEvents
+    Supports both binary and structured mode CloudEvents
     """
 
+    @classmethod
+    def from_http(
+        cls,
+        data: typing.Union[str, bytes],
+        headers: typing.Dict[str, str],
+        data_unmarshaller: types.UnmarshallerType = None
+    ):
+        """Unwrap a CloudEvent (binary or structured) from an HTTP request.
+        :param data: the HTTP request body
+        :type data: typing.IO
+        :param headers: the HTTP headers
+        :type headers: typing.Dict[str, str]
+        :param data_unmarshaller: Function to decode data into a python object.
+        :type data_unmarshaller: types.UnmarshallerType
+        """
+        if data_unmarshaller is None:
+            data_unmarshaller = _json_or_string
+
+        event = marshaller.NewDefaultHTTPMarshaller().FromRequest(
+            v1.Event(), headers, data, data_unmarshaller)
```

Comment:
Until we've parsed the event, it's hard to find the specversion. Fortunately, since we extract everything from the Event and throw it away, all we need are the parsing, which seems to be the same path for v3 and v1.

I agree that it's not ideal, but I think it will take a bunch of refactoring to separate the attribute extraction from the representation.

Reply:
Alright. I'll make an issue for this once the PR gets merged.